src/main/java/com/microsoft/azure/util/AzureMsiCredentials.java

```diff
@@ -2,24 +2,87 @@
 
 import com.cloudbees.plugins.credentials.CredentialsScope;
 import com.cloudbees.plugins.credentials.impl.BaseStandardCredentials;
+import com.microsoft.azure.AzureEnvironment;
 import hudson.Extension;
+import hudson.util.ListBoxModel;
+import org.apache.commons.lang.StringUtils;
 import org.kohsuke.stapler.DataBoundConstructor;
 
+import java.util.Map;
+
 public class AzureMsiCredentials extends BaseStandardCredentials {
 
     public static final int DEFAULT_MSI_PORT = 50342;
+    private static final long serialVersionUID = 6743945878507124459L;
+
     private final int msiPort;
+    private String azureEnvName;
+    private transient AzureEnvironment azureEnvironment;
 
     @DataBoundConstructor
-    public AzureMsiCredentials(CredentialsScope scope, String id, String description, int msiPort) {
+    public AzureMsiCredentials(CredentialsScope scope, String id, String description, int msiPort,
+                               String azureEnvName) {

```

Comment:
This may cause compatibility issue if we have dependency on the previous constructor signature, although currently the usage there may be no such dependency.

I suggest we use `@DataBoundSetter` instead of `@DataBoundConstructor`, which offers more flexibility.

Reply:
`azureEnvName` is a required parameter, so I believe it's more sensible to put it in the constructor

Reply:
As for the current implementation, `azureEnvName` can be ignored and we will fallback to the default `Azure`. This is how we ensure the backward compatibility of the existing credentials configuration.

Strictly speaking, this is a breaking change and we need to advance the major version. If you need this constructor for sure, please add back the original constructor and mark it as `@deprecated`.

Thanks,